cmd/gaia/app/app.go

```diff
@@ -0,0 +1,200 @@
+package app
+
+import (
+	"encoding/json"
+
+	abci "github.com/tendermint/abci/types"
+	cmn "github.com/tendermint/tmlibs/common"
+	dbm "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+
+	bam "github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/wire"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+	"github.com/cosmos/cosmos-sdk/x/ibc"
+	"github.com/cosmos/cosmos-sdk/x/stake"
+)
+
+const (
+	appName = "GaiaApp"
+)
+
+// Extended ABCI application
+type GaiaApp struct {
+	*bam.BaseApp
+	cdc *wire.Codec
+
+	// keys to access the substores
+	capKeyMainStore    *sdk.KVStoreKey
+	capKeyAccountStore *sdk.KVStoreKey
+	capKeyIBCStore     *sdk.KVStoreKey
+	capKeyStakeStore   *sdk.KVStoreKey
+
+	// Manage getting and setting accounts
+	accountMapper sdk.AccountMapper
+	coinKeeper    bank.CoinKeeper
+	ibcMapper     ibc.IBCMapper
+	stakeKeeper   stake.Keeper
+
+	// Handle fees
+	feeHandler sdk.FeeHandler
+}
+
+func NewGaiaApp(logger log.Logger, db dbm.DB) *GaiaApp {
+	// create your application object
+	var app = &GaiaApp{
+		BaseApp:            bam.NewBaseApp(appName, logger, db),
+		cdc:                MakeCodec(),
+		capKeyMainStore:    sdk.NewKVStoreKey("main"),
+		capKeyAccountStore: sdk.NewKVStoreKey("acc"),
+		capKeyIBCStore:     sdk.NewKVStoreKey("ibc"),
+		capKeyStakeStore:   sdk.NewKVStoreKey("stake"),
+	}
+
+	// define the accountMapper
+	app.accountMapper = auth.NewAccountMapper(
+		app.cdc,
+		app.capKeyMainStore, // target store
+		&auth.BaseAccount{}, // prototype
+	).Seal()
+
+	// add handlers
+	app.coinKeeper = bank.NewCoinKeeper(app.accountMapper)
+	app.ibcMapper = ibc.NewIBCMapper(app.cdc, app.capKeyIBCStore, app.RegisterCodespace(ibc.DefaultCodespace))
+	app.stakeKeeper = stake.NewKeeper(app.cdc, app.capKeyStakeStore, app.coinKeeper, app.RegisterCodespace(stake.DefaultCodespace))
+
+	app.Router().
+		AddRoute("bank", bank.NewHandler(app.coinKeeper)).
+		AddRoute("ibc", ibc.NewHandler(app.ibcMapper, app.coinKeeper)).
+		AddRoute("stake", stake.NewHandler(app.stakeKeeper))
+
+	// Define the feeHandler.
+	app.feeHandler = auth.BurnFeeHandler
+
+	// initialize BaseApp
+	app.SetTxDecoder(app.txDecoder)
+	app.SetInitChainer(app.initChainer)
+	app.SetEndBlocker(stake.NewEndBlocker(app.stakeKeeper))
+	app.MountStoresIAVL(app.capKeyMainStore, app.capKeyAccountStore, app.capKeyIBCStore, app.capKeyStakeStore)
+	app.SetAnteHandler(auth.NewAnteHandler(app.accountMapper, app.feeHandler))
+	err := app.LoadLatestVersion(app.capKeyMainStore)
+	if err != nil {
+		cmn.Exit(err.Error())
+	}
+
+	return app
+}
+
+// custom tx codec
+func MakeCodec() *wire.Codec {
+	var cdc = wire.NewCodec()
+
+	// Register Msgs
+	cdc.RegisterInterface((*sdk.Msg)(nil), nil)
+
+	ibc.RegisterWire(cdc)
```

Comment:
This is great, we should update `basecoin` / `democoin` to use the module `RegisterWire` functions too (doesn't need to happen in this PR).

Reply:
Let's do a seperate PR for that - I'll lump it with a cleanup PR I'd like to merge today